### PR TITLE
test: reduce testing to 2 python versions per sample

### PR DESCRIPTION
**Pull request by parthea**
Please review https://github.com/GoogleCloudPlatform/python-docs-samples/pull/8872 first

Once https://github.com/GoogleCloudPlatform/python-docs-samples/pull/8872 is merged, the versions of python tested will be 2.7, 3.7, 3.8, 3.9, 3.10 and 3.11

To save on resources, let's restrict testing to the end points. So, 3.7 and 3.11 (plus 2.7 when needed).

### CODE REVIEW

**Comment by leahecole** [comment]
@parthea I know that Composer is an edge case that for now only is in Python 3.8. How should we handle such cases? otherwise I think this is a great idea

**Comment by leahecole** [comment]
removed earlier comment, realized we were editing the noxfile configs per product not the base test image. sorry!

**Comment by ace-n** [comment]
FYI: App Engine [does not fully support 3.11 yet](https://cloud.google.com/appengine/docs/flexible/python/runtime#:~:text=App%20Engine%20standard%20environment%3A%20Python,3.9%2C%20and%203.10%20are%20supported.).

**Comment by leahecole** [comment]
Composer LGTM
data-science-onramp - I am seeing the protocol buffers descriptors error (but don't with Python 3.10 or below) - I am wondering if it's really just that python-aiplatform isn't ready yet?

**Comment by leahecole** [comment]
I'm going to merge with main to kick off another build but I'm thinking that...

3.7
- retail failure looks like it might be resolved by #8968 

3.11
- dataproc should be fixed now that I gave the default service account permissions
- talent/jobs should be resolved from enabling the api and approving the log thing
- healthcare should be resolved since we enabled the api
- secretmanager should be resolved since we enabled the API

**Comment by dandhlee** [comment]
I believe some products only serve 3.8 or 3.9 (Run and Data* products).

Fundamentally, I'd strongly push back against making this change. If users run into problems before we do, simply because we reduced the number of versions being tested, that'd be a really bad experience for everyone.

If we want to save resources, can we run the nightly / continuous less frequently instead?

**Comment by kweinmeister** [comment]
@dandhlee I believe this change does not affect samples that have a dependency on a specific version (e.g. 3.8 or 3.9). It should only address projects with open-ended 3.x support. If there are spots where the PR is not meeting this criteria, we should definitely update it.

This PR is aligned with a risk-based testing strategy that is followed in other languages (e.g. Java 8, 11, 17 and not 8..17). If the endpoints are both passing, the risk of a deterministic language-specific issue in 3.8..3.10 is very low and acceptable for sample code given tradeoffs IMO. We can incrementally add versions back to the matrix if such issues arise -- e.g. a midpoint version (3.9) next. Also, in addition to this change, as you suggested we can review the frequency of periodic builds.

FYI @nicain

**Comment by leahecole** [comment]
> FYI: App Engine [does not fully support 3.11 yet](https://cloud.google.com/appengine/docs/flexible/python/runtime#:~:text=App%20Engine%20standard%20environment%3A%20Python,3.9%2C%20and%203.10%20are%20supported.).

@ace-n the doc you linked is for app engine standard, but iiuc these tests are for app engine flexible - does flexible also not support 3.11?

**Comment by ace-n** [comment]
@leahecole here's my understanding:

### GAE Flex
Flexible [supports](https://cloud.google.com/appengine/docs/flexible/python/runtime) Python 3.4 - 3.7.

### GAE Standard
The App Engine standard template (`appengine/standard/noxfile-template.py`)'s `TEST_CONFIG["ignored_versions"]` **does not match** the `IGNORED_VERSIONS` variable lower in the file, and I'm not sure which one takes priority.

### GAE `standard_python3`
For `appengine/standard_python3` - this PR would test `3.7` and `3.11`. I assume we should change `3.11` to `3.10`? (cc @engelke as FYI)

File: appengine/standard_python3/bundled-services/blobstore/django/noxfile_config.py [hunk 1]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/blobstore/django/noxfile_config.py:28]:
```suggestion
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: appengine/standard_python3/bundled-services/blobstore/flask/noxfile_config.py [hunk 2]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/blobstore/flask/noxfile_config.py:28]:
```suggestion
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: appengine/standard_python3/bundled-services/blobstore/wsgi/noxfile_config.py [hunk 3]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/blobstore/wsgi/noxfile_config.py:28]:
```suggestion
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: appengine/standard_python3/bundled-services/deferred/django/noxfile_config.py [hunk 4]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/deferred/django/noxfile_config.py:28]:
```suggestion
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: appengine/standard_python3/bundled-services/deferred/flask/noxfile_config.py [hunk 5]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/deferred/flask/noxfile_config.py:28]:
```suggestion
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: appengine/standard_python3/bundled-services/deferred/wsgi/noxfile_config.py [hunk 6]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/deferred/wsgi/noxfile_config.py:28]:
```suggestion
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: appengine/standard_python3/bundled-services/mail/django/noxfile_config.py [hunk 7]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/mail/django/noxfile_config.py:28]:
```suggestion
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: appengine/standard_python3/bundled-services/deferred/django/noxfile_config.py [hunk 8]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/deferred/django/noxfile_config.py:26]:
```suggestion
    # Test min and max supported versions which are python 3.7 and 3.10
```

File: appengine/standard_python3/bundled-services/deferred/flask/noxfile_config.py [hunk 9]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/deferred/flask/noxfile_config.py:26]:
```suggestion
    # Test min and max supported versions which are python 3.7 and 3.10
```

File: appengine/standard_python3/bundled-services/deferred/wsgi/noxfile_config.py [hunk 10]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/deferred/wsgi/noxfile_config.py:26]:
```suggestion
    # Test min and max supported versions which are python 3.7 and 3.10
```

File: appengine/standard_python3/bundled-services/mail/django/noxfile_config.py [hunk 11]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/mail/django/noxfile_config.py:26]:
```suggestion
    # Test min and max supported versions which are python 3.7 and 3.10
```

File: appengine/standard_python3/bundled-services/mail/flask/noxfile_config.py [hunk 12]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/mail/flask/noxfile_config.py:28]:
```suggestion
    # Test min and max supported versions which are python 3.7 and 3.10
    # Ignoring python 2.7 which is no longer supported.
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: appengine/standard_python3/bundled-services/mail/wsgi/noxfile_config.py [hunk 13]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7", "3.6"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/bundled-services/mail/wsgi/noxfile_config.py:28]:
```suggestion
    # Test min and max supported versions which are python 3.7 and 3.10
    # Ignoring python 2.7 which is no longer supported.
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: appengine/standard_python3/django/noxfile_config.py [hunk 14]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    'ignored_versions': ["2.7"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline appengine/standard_python3/django/noxfile_config.py:28]:
```suggestion
    # Test min and max supported versions which are python 3.7 and 3.10
    # Ignoring python 2.7 which is no longer supported.
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

File: asset/snippets/noxfile_config.py [hunk 15]
```diff
@@ -21,8 +21,11 @@
 # https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/noxfile_config.py
 
 TEST_CONFIG_OVERRIDE = {
-    # You can opt out from the test for specific Python versions.
-    "ignored_versions": ["2.7"],
+    # You can opt out specific Python versions from the test.
+    # To save resources, please consider testing 2 versions at most.
+    # Test min and max supported versions which are python 3.7 and 3.11
+    # Ignoring python 2.7 which is no longer supported.
+    "ignored_versions": ["2.7", "3.8", "3.9", "3.10"],
```
Comment (leahecole) [inline asset/snippets/noxfile_config.py:28]:
```suggestion
    # Test min and max supported versions which are python 3.7 and 3.10
    # Ignoring python 2.7 which is no longer supported.
    "ignored_versions": ["2.7", "3.8", "3.9", "3.11"],
```

**Review by leahecole — CHANGES_REQUESTED** [review]
requesting changes to appengine per @ace-n 's comment

**Comment by engelke** [comment]
A lot has changed since this PR was proposed, and there are a lot of items for discussion. Could we close this for now and revisit the whole thing following some planned fixes?

**Pull request closed without merging**